File: bindings/python/tests/test_submit_result_terminal_contract.py
import asyncio
import uuid

import pytest
import zlink


HWM_BYTES = 512
MAX_FILL_RECORDS = 4096
PAYLOAD_BYTES = 64


def _payload(sequence):
    return sequence.to_bytes(4, "little") + b"z" * (PAYLOAD_BYTES - 4)


def _configure_small_hwm(socket):
    socket.options.linger_ms = 0
    socket.options.send_high_water_mark = HWM_BYTES
    socket.options.receive_high_water_mark = HWM_BYTES


def _connect_ready(server, client, label):
    endpoint = f"inproc://python-{label}-{uuid.uuid4().hex}"
    server.bind(endpoint)
    client.connect(endpoint)
    client.send().message(b"ready").submit_sync()
    received = zlink.create_received()
    try:
        assert server.recv_into(received)
        assert received.to_bytes_list() == [b"ready"]
    finally:
        received.close()


def _receive(server):
    received = zlink.create_received()
    assert server.recv_into(received)
    return received


def _poll_until(poller, predicate):
    events = zlink.create_poll_events(4)
    for _ in range(32):
        if predicate():
            return
        assert poller.wait(events, 5000) > 0
    assert predicate()


@pytest.mark.parametrize("round_index", range(5))
def test_immediate_admission_returns_ok_with_completed_stage_before_reply(
    round_index,
):
    async def exercise():
        with zlink.create_context() as context:
            with zlink.create_router_socket(context) as server:
                with zlink.create_dealer_socket(context) as client:
                    _connect_ready(server, client, f"submit-ok-{round_index}")

                    send = client.send().message(b"send-ok").submit()
                    assert isinstance(send, zlink.SendSubmission)
                    assert send.result == zlink.SubmitResult.OK
                    assert send.admitted.done()
                    assert await send.admitted is None
                    received = _receive(server)
                    try:
                        assert received.to_bytes_list() == [b"send-ok"]
                    finally:
                        received.close()

                    request = (
                        client.request().message(b"request-ok").timeout(30).submit()
                    )
                    assert isinstance(request, zlink.RequestSubmission)
                    assert request.result == zlink.SubmitResult.OK
                    assert request.admitted.done()
                    assert await request.admitted is None
                    assert not request.reply.done()

                    received = _receive(server)
                    try:
                        assert received.to_bytes_list() == [b"request-ok"]
                        received.reply().message(b"reply-ok").submit()
                    finally:
                        received.close()
                    reply = await asyncio.wait_for(
                        asyncio.shield(request.reply), 5
                    )
                    try:
                        assert [part.to_bytes() for part in reply] == [b"reply-ok"]
                    finally:
                        for part in reply:
                            part.close()

    asyncio.run(exercise())


@pytest.mark.parametrize("round_index", range(5))
def test_hwm_backpressure_admits_after_public_writable_progress_before_reply(
    round_index,
):
    async def exercise():
        with zlink.create_context() as context:
            context.options.auto_hwm_enabled = False
            with zlink.create_router_socket(context) as server:
                with zlink.create_dealer_socket(context) as client:
                    _configure_small_hwm(server)
                    _configure_small_hwm(client)
                    _connect_ready(server, client, f"submit-hwm-{round_index}")

                    with zlink.create_poller() as poller:
                        poller.add_socket(
                            client,
                            zlink.PollEventFlag.POLLOUT
                            | zlink.PollEventFlag.POLLCOMPLETION,
                            1,
                        )

                        admitted_sends = 0
                        waiting_send = None
                        for sequence in range(MAX_FILL_RECORDS):
                            submission = (
                                client.send().message(_payload(sequence)).submit()
                            )
                            if submission.result == zlink.SubmitResult.BACKPRESSURED:
                                waiting_send = submission
                                break
                            assert submission.result == zlink.SubmitResult.OK
                            assert submission.admitted.done()
                            admitted_sends += 1
                        assert admitted_sends > 0
                        assert waiting_send is not None
                        assert not waiting_send.admitted.done()

                        for sequence in range(admitted_sends):
                            received = _receive(server)
                            try:
                                assert received.to_bytes_list() == [_payload(sequence)]
                            finally:
                                received.close()
                        _poll_until(poller, waiting_send.admitted.done)
                        assert await waiting_send.admitted is None
                        assert waiting_send.result == zlink.SubmitResult.BACKPRESSURED
                        received = _receive(server)
                        try:
                            assert received.to_bytes_list() == [
                                _payload(admitted_sends)
                            ]
                        finally:
                            received.close()

                        admitted_requests = []
                        waiting_request = None
                        for sequence in range(MAX_FILL_RECORDS):
                            submission = (
                                client.request()
                                .message(_payload(sequence))
                                .timeout(30)
                                .submit()
                            )
                            if submission.result == zlink.SubmitResult.BACKPRESSURED:
                                waiting_request = submission
                                break
                            assert submission.result == zlink.SubmitResult.OK
                            assert submission.admitted.done()
                            admitted_requests.append(submission)
                        assert admitted_requests
                        assert waiting_request is not None
                        assert not waiting_request.admitted.done()
                        assert not waiting_request.reply.done()

                        for sequence in range(len(admitted_requests)):
                            received = _receive(server)
                            try:
                                assert received.to_bytes_list() == [
                                    _payload(sequence)
                                ]
                                received.reply().message(
                                    b"reply-" + sequence.to_bytes(4, "little")
                                ).submit()
                            finally:
                                received.close()
                        _poll_until(poller, waiting_request.admitted.done)
                        assert await waiting_request.admitted is None
                        assert (
                            waiting_request.result
                            == zlink.SubmitResult.BACKPRESSURED
                        )
                        assert not waiting_request.reply.done()

                        received = _receive(server)
                        try:
                            assert received.to_bytes_list() == [
                                _payload(len(admitted_requests))
                            ]
                            received.reply().message(b"reply-waiting").submit()
                        finally:
                            received.close()
                        _poll_until(poller, waiting_request.reply.done)

                        replies = await asyncio.gather(
                            *(submission.reply for submission in admitted_requests),
                            waiting_request.reply,
                        )
                        try:
                            assert replies[-1][0].to_bytes() == b"reply-waiting"
                        finally:
                            for parts in replies:
                                for part in parts:
                                    part.close()

    asyncio.run(exercise())

File: bindings/dotnet/tests/Zlink.Tests/test_submit_result_terminal_contract.cs
using Xunit;

namespace Systems.Zlink.Tests;

public sealed class test_submit_result_terminal_contract
{
    private const ulong RecordHwm = 65_536UL + 64UL;
    private const int MaxFillRecords = 32;
    private static readonly string FillerPayload =
        "submit-result-" + new string('z', 65_536);

    [Theory]
    [InlineData(1)]
    [InlineData(2)]
    [InlineData(3)]
    [InlineData(4)]
    [InlineData(5)]
    public async Task immediate_admission_returns_ok_before_request_reply(
        int repetition)
    {
        if (!CoreTestSupport.IsNativeAvailable())
            return;

        using var context = Zlink.CreateContext();
        using var server = context.CreateRouterSocket();
        using var client = context.CreateDealerSocket();
        ConnectReady(server, client, $"submit-result-ok-{repetition}");

        using (Message payload = Message.From("send-ok"))
        {
            SendSubmission send = client.Send().Message(payload).Async();
            Assert.Equal(SubmitResult.Ok, send.Result);
            Assert.True(send.Admitted.IsCompletedSuccessfully);
            await send.Admitted;
        }
        AssertReceived(server, "send-ok");

        RequestSubmission request;
        using (Message payload = Message.From("request-ok"))
        {
            request = client.Request().Message(payload)
                .Timeout(TimeSpan.FromSeconds(30)).Async();
        }
        Assert.Equal(SubmitResult.Ok, request.Result);
        Assert.True(request.Admitted.IsCompletedSuccessfully);
        Assert.False(request.Reply.IsCompleted);

        ReplyOnce(server, "request-ok", "reply-ok");
        IReadOnlyList<Message> reply = await request.Reply.WaitAsync(
            TimeSpan.FromSeconds(5));
        try
        {
            Assert.Equal("reply-ok", Assert.Single(reply).GetString());
        }
        finally
        {
            Zlink.MultipartClose(reply);
        }
    }

    [Theory]
    [InlineData(1)]
    [InlineData(2)]
    [InlineData(3)]
    [InlineData(4)]
    [InlineData(5)]
    public async Task hwm_returns_backpressured_then_admits_before_reply(
        int repetition)
    {
        if (!CoreTestSupport.IsNativeAvailable())
            return;

        await VerifyBackpressuredSend(repetition);
        await VerifyBackpressuredRequest(repetition);
    }

    private static async Task VerifyBackpressuredSend(int repetition)
    {
        using var context = Zlink.CreateContext();
        context.Options.AutoHwmEnabled = false;
        using var server = context.CreateRouterSocket();
        using var client = context.CreateDealerSocket();
        using var poller = Zlink.CreatePoller();
        ConfigureSmallHwm(server, client);
        ConnectReady(server, client, $"submit-result-send-hwm-{repetition}");
        poller.Add(client,
            PollEventFlags.PollOut | PollEventFlags.PollCompletion, 1);

        SendSubmission waiting = default;
        var acceptedCount = 0;
        for (var sequence = 0; sequence < MaxFillRecords; sequence++)
        {
            using Message payload = Message.From(Payload(sequence));
            SendSubmission submission = client.Send().Message(payload).Async();
            if (submission.Result == SubmitResult.Backpressured)
            {
                waiting = submission;
                break;
            }
            Assert.Equal(SubmitResult.Ok, submission.Result);
            Assert.True(submission.Admitted.IsCompletedSuccessfully);
            acceptedCount++;
        }

        Assert.True(acceptedCount > 0);
        Assert.Equal(SubmitResult.Backpressured, waiting.Result);
        Assert.False(waiting.Admitted.IsCompleted);
        Assert.Equal(0, poller.Wait(new PollEvent[1], TimeSpan.Zero));

        for (var sequence = 0; sequence < acceptedCount; sequence++)
            AssertReceived(server, Payload(sequence));

        AwaitPollUntil(poller, waiting.Admitted);
        await waiting.Admitted.WaitAsync(TimeSpan.FromSeconds(5));
        Assert.Equal(SubmitResult.Backpressured, waiting.Result);
        AssertReceived(server, Payload(acceptedCount));
    }

    private static async Task VerifyBackpressuredRequest(int repetition)
    {
        using var context = Zlink.CreateContext();
        context.Options.AutoHwmEnabled = false;
        using var server = context.CreateRouterSocket();
        using var client = context.CreateDealerSocket();
        using var poller = Zlink.CreatePoller();
        ConfigureSmallHwm(server, client);
        ConnectReady(server, client,
            $"submit-result-request-hwm-{repetition}");
        poller.Add(client,
            PollEventFlags.PollOut | PollEventFlags.PollCompletion, 1);

        var admitted = new List<RequestSubmission>();
        RequestSubmission waiting = default;
        for (var sequence = 0; sequence < MaxFillRecords; sequence++)
        {
            using Message payload = Message.From(Payload(sequence));
            RequestSubmission submission = client.Request().Message(payload)
                .Timeout(TimeSpan.FromSeconds(30)).Async();
            if (submission.Result == SubmitResult.Backpressured)
            {
                waiting = submission;
                break;
            }
            Assert.Equal(SubmitResult.Ok, submission.Result);
            Assert.True(submission.Admitted.IsCompletedSuccessfully);
            admitted.Add(submission);
        }

        Assert.NotEmpty(admitted);
        Assert.Equal(SubmitResult.Backpressured, waiting.Result);
        Assert.False(waiting.Admitted.IsCompleted);
        Assert.False(waiting.Reply.IsCompleted);
        Assert.Equal(0, poller.Wait(new PollEvent[1], TimeSpan.Zero));

        for (var sequence = 0; sequence < admitted.Count; sequence++)
            ReplyOnce(server, Payload(sequence), $"reply-{sequence}");

        AwaitPollUntil(poller, waiting.Admitted);
        await waiting.Admitted.WaitAsync(TimeSpan.FromSeconds(5));
        Assert.Equal(SubmitResult.Backpressured, waiting.Result);
        Assert.False(waiting.Reply.IsCompleted);

        ReplyOnce(server, Payload(admitted.Count), "reply-waiting");
        AwaitPollUntil(poller, waiting.Reply);
        for (var sequence = 0; sequence < admitted.Count; sequence++)
        {
            IReadOnlyList<Message> reply = await admitted[sequence].Reply
                .WaitAsync(TimeSpan.FromSeconds(5));
            try
            {
                Assert.Equal($"reply-{sequence}",
                    Assert.Single(reply).GetString());
            }
            finally
            {
                Zlink.MultipartClose(reply);
            }
        }

        IReadOnlyList<Message> waitingReply = await waiting.Reply.WaitAsync(
            TimeSpan.FromSeconds(5));
        try
        {
            Assert.Equal("reply-waiting",
                Assert.Single(waitingReply).GetString());
        }
        finally
        {
            Zlink.MultipartClose(waitingReply);
        }
    }

    private static void ConnectReady(IRouterSocket server,
        IDealerSocket client, string label)
    {
        string endpoint = CoreTestSupport.NewEndpoint("inproc", label);
        server.Bind(endpoint);
        client.Connect(endpoint);
        using Message probe = Message.From("ready");
        client.Send().Message(probe).Submit();
        AssertReceived(server, "ready");
    }

    private static void ConfigureSmallHwm(IRouterSocket server,
        IDealerSocket client)
    {
        server.Options.Linger = TimeSpan.Zero;
        client.Options.Linger = TimeSpan.Zero;
        server.Options.ReceiveHighWaterMark = RecordHwm;
        client.Options.SendHighWaterMark = RecordHwm;
    }

    private static void AwaitPollUntil(IPoller poller, Task completion)
    {
        var events = new PollEvent[1];
        for (var attempt = 0; attempt < 32 && !completion.IsCompleted;
             attempt++)
        {
            Assert.Equal(1, poller.Wait(events, TimeSpan.FromSeconds(5)));
            Assert.Equal((nuint)1, events[0].Slot);
        }
        Assert.True(completion.IsCompleted);
    }

    private static void ReplyOnce(IRouterSocket server, string expected,
        string reply)
    {
        using Received received = Received.Create();
        Assert.True(server.Recv(received));
        Assert.Equal(expected,
            received.SinglePartOrThrow().GetString());
        using Message payload = Message.From(reply);
        received.Reply().Message(payload).Submit();
    }

    private static void AssertReceived(IRouterSocket server, string expected)
    {
        using Received received = Received.Create();
        Assert.True(server.Recv(received));
        Assert.Equal(expected,
            received.SinglePartOrThrow().GetString());
    }

    private static string Payload(int sequence) =>
        FillerPayload + $"-{sequence:D2}";
}
